src/Control/Carrier/Accum/Church.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- | A high-performance, strict, church-encoded carrier for 'Accum'.
+
+This carrier issues left-associated 'mappend's, meaning that 'Monoid's such as @[]@ with poor performance for left-associated 'mappend's are ill-suited for use with this carrier. Alternatives such as 'Data.Monoid.Endo', @Seq@, or @DList@ may be preferred.
+
+-- | @since 1.0.3.0 -- FIXME Is this right?
+-}
+
+module Control.Carrier.Accum.Church
+( -- * Accum carrier
+  runAccum
+, execAccum
+, AccumC(..)
+  -- * Accum effect
+, module Control.Effect.Accum
+) where
+
+import Control.Algebra
+import Control.Applicative (Alternative)
+import Control.Carrier.State.Church
+import Control.Effect.Accum
+import Control.Monad (MonadPlus)
+import Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+
+-- | Run an 'Accum' effect with a 'Monoid'al log, applying a continuation to the final log and result.
+--
+-- @
+-- 'runAccum' k ('pure' a) = k 'mempty' a
+-- @
+-- @
+-- 'runAccum' k ('add' w) = k w ()
+-- @
+-- @
+-- 'runAccum' k ('add' w >> 'look') = k w w
+-- @
+--
+-- @since 1.0.3.0 -- FIXME Is this right?
+runAccum :: Monoid w => (w -> a -> m b) -> AccumC w m a -> m b
+runAccum k = runState k mempty . runAccumC
+{-# INLINE runAccum #-}
+
+-- | Run a 'Accum' effect with a 'Monoid'al log, producing the final log and discarding the result value.
+--
+-- @
+-- 'execAccum' = 'runAccum' ('const' '.' 'pure')
+-- @
+--
+-- @since 1.0.3.0 -- FIXME Is this right?
+execAccum :: (Monoid w, Applicative m) => AccumC w m a -> m w
+execAccum = runAccum (const . pure)
+{-# INLINE execAccum #-}
```

Comment:
This is lovely! Might I request an `evalAccum` as well (by analogy with `evalState` &c.)?